### PR TITLE
Add Quantinuum support for Python SDK

### PR DESCRIPTION
**Pull request by vxfield**
The goal of this PR is to add support for a new `Quantinuum` provider and corresponding targets, while still keeping the `Honeywell` providers and targets fully functional.

In general, that's accomplished by new `Quantinuum` classes and methods that inherits/calls existing `Honeywell` classes/methods.
When needed, an optional `provider_id` parameter was introduced to improve code reusability of the existing classes/methods.

### CODE REVIEW

File: azure-quantum/azure/quantum/aio/target/quantinuum.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+##
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (cgranade) [inline azure-quantum/azure/quantum/aio/target/quantinuum.py:2]:
```suggestion
# Copyright (c) Microsoft Corporation.
```

File: azure-quantum/azure/quantum/aio/target/quantinuum.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    async def submit(
+        self,
+        circuit: str,
+        name: str = "quantinuum-job",
+        num_shots: int = None,
+        input_params: Dict[str, Any] = None,
+        **kwargs
+    ) -> Job:
+        """Submit a Quantinuum program (QASM format)
```
Comment (cgranade) [inline azure-quantum/azure/quantum/aio/target/quantinuum.py:23]: QASM on its own is ambiguous, as it could refer to OpenQASM 2.0, OpenQASM 3.0, cQASM, eQASM, or even the older qasm2circ format.

```suggestion
        """Submit a Quantinuum program (OpenQASM 2.0 format)
```

File: azure-quantum/azure/quantum/aio/target/quantinuum.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+        :param circuit: Quantum circuit in Quantinuum (formerly Honeywell) QASM format
```
Comment (cgranade) [inline azure-quantum/azure/quantum/aio/target/quantinuum.py:25]: Similarly here.

File: azure-quantum/tests/unit/common.py [hunk 4]
```diff
@@ -57,6 +58,9 @@ def __init__(self, method_name):
         self._location = os.environ.get("AZURE_QUANTUM_WORKSPACE_LOCATION", os.environ.get("LOCATION", LOCATION))
         self._user_agent = os.environ.get("AZURE_QUANTUM_PYTHON_APPID")
 
+        # needed to make sure the Default Credential will work
+        os.environ["LOCATION"] = self._location
```
Comment (cgranade) [inline azure-quantum/tests/unit/common.py:62]: Is this change related?

File: azure-quantum/tests/unit/common.py [hunk 5]
```diff
@@ -75,6 +79,7 @@ def __init__(self, method_name):
             OAuthRequestResponsesFilter(),
             RequestUrlNormalizer(),
             OAuthResponsesFilter(),
+            CustomUrlPlaybackProcessor()
```
Comment (cgranade) [inline azure-quantum/tests/unit/common.py:82]: Similarly here, this seems unrelated to the rebranding?

File: azure-quantum/tests/unit/test_qiskit.py [hunk 6]
```diff
@@ -41,10 +41,9 @@ def _3_qubit_ghz(self):
         circuit.cx(0, 1)
         circuit.cx(1, 2)
         circuit.h(3) # Dummy helper qubit that is not measured
```
Comment (cgranade) [inline azure-quantum/tests/unit/test_qiskit.py:43]:
```suggestion
        circuit.h(3) # Helper qubit that is not measured
```

**Review by cgranade — APPROVED** [review]

File: azure-quantum/tests/unit/common.py (hunk 4 above)
Comment (vxfield) [reply azure-quantum/tests/unit/common.py:62]: Not directly related, but helps with the recording of unit tests following the readme.
Since it's just a unit test, it has no effect on the release and I would like to keep it.

File: azure-quantum/azure/quantum/aio/target/quantinuum.py (hunk 3 above)
Comment (vxfield) [reply azure-quantum/azure/quantum/aio/target/quantinuum.py]: I replaced the documentation in other places too.

File: azure-quantum/azure/quantum/cirq/targets/honeywell.py [hunk 7]
```diff
@@ -17,7 +17,7 @@
 
 
 class HoneywellTarget(Honeywell, CirqTarget):
```
Comment (guenp) [inline azure-quantum/azure/quantum/cirq/targets/honeywell.py:19]: I suggest renaming this to `QuantinuumTarget` and adding a new subclass `HoneywellTarget` that has a deprecation warning in the constructor

File: azure-quantum/azure/quantum/aio/target/quantinuum.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+# Licensed under the MIT License.
+##
+from typing import Any, Dict
+
+from azure.quantum.aio.target.target import Target
+from azure.quantum.aio.job.job import Job
+from azure.quantum.aio.target import Honeywell
+
+
+class Quantinuum(Honeywell):
```
Comment (guenp) [inline azure-quantum/azure/quantum/aio/target/quantinuum.py:12]: I think we should do this the other way around so we can easily deprecate Honeywell.
I suggest renaming `Honeywell` to `Quantinuum` and adding a new subclass `Honeywell` that has a deprecation warning in the constructor

File: azure-quantum/azure/quantum/qiskit/backends/honeywell.py [hunk 9]
```diff
@@ -56,9 +56,18 @@
     "reset"
 ]
 
+QUANTINUUM_PROVIDER_NAMES = {"honeywell": "Honeywell", "quantinuum": "Quantinuum"}
```
Comment (guenp) [inline azure-quantum/azure/quantum/qiskit/backends/honeywell.py:59]: I suggest to make this consistent with the rest of the code
```suggestion
QUANTINUUM_PROVIDER_ID = "quantinuum"
HONEYWELL_PROVIDER_ID = "honeywell"
```

File: azure-quantum/azure/quantum/qiskit/backends/honeywell.py [hunk 10]
```diff
@@ -223,14 +244,19 @@ def __init__(
...
+
         default_config = BackendConfiguration.from_dict(
             {
                 "backend_name": name,
                 "backend_version": __version__,
                 "simulator": False,
                 "local": False,
                 "coupling_map": None,
-                "description": "Honeywell QPU on Azure Quantum",
+                "description": f"{self._provider_name} QPU on Azure Quantum",
```
Comment (guenp) [inline azure-quantum/azure/quantum/qiskit/backends/honeywell.py:259]:
```suggestion
                "description": f"{HONEYWELL_PROVIDER_ID} QPU on Azure Quantum",
```

File: azure-quantum/azure/quantum/qiskit/backends/honeywell.py [hunk 11]
```diff
@@ -223,14 +244,19 @@ def __init__(
         provider: "AzureQuantumProvider",
         **kwargs
     ):
+        self._provider_id = "honeywell"
+        if "provider_id" in kwargs:
+            self._provider_id = kwargs.pop("provider_id")
+        self._provider_name = QUANTINUUM_PROVIDER_NAMES[self._provider_id]
```
Comment (guenp) [inline azure-quantum/azure/quantum/qiskit/backends/honeywell.py:250]:
```suggestion
```

File: azure-quantum/azure/quantum/qiskit/backends/honeywell.py [hunk 12]
```diff
@@ -242,5 +268,5 @@ def __init__(
             }
         )
         configuration: BackendConfiguration = kwargs.pop("configuration", default_config)
-        logger.info("Initializing HoneywellQPUBackend")
+        logger.info(f"Initializing {self._provider_name}QPUBackend")
```
Comment (guenp) [inline azure-quantum/azure/quantum/qiskit/backends/honeywell.py:271]:
```suggestion
        logger.info(f"Initializing {HONEYWELL_PROVIDER_ID}QPUBackend")
```

**Review by guenp — CHANGES_REQUESTED** [review]

**Comment by vxfield** [comment]
Thanks, @guenp. I resolved all comments as I made the suggested changes in the code.

**Review by guenp — APPROVED** [review]
Thanks for making those changes, LGTM!
